### PR TITLE
EWPP-453: Execute and render the RSS list

### PR DESCRIPTION
**Pull request by imanoleguskiza**
### Description

Execute and render the RSS list for list pages.

### Change log

- Added:
- Changed: Execute and render the RSS list
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: src/Controller/ListPageRssController.php [hunk 1]
```diff
@@ -164,12 +191,44 @@ public function build(NodeInterface $node): CacheableResponse {
...
+      $cache_metadata->addCacheableDependency($entity);
+      $creation_date = $entity->get('changed')->value;
+      $result_item = [
+        '#theme' => 'oe_list_pages_rss_item',
+        '#title' => $entity->label(),
+        '#link' => $entity->toUrl('canonical', ['absolute' => TRUE]),
+        '#guid' => $entity->toUrl('canonical', ['absolute' => TRUE]),
+        '#description' => '',
+        '#item_elements' => [
+          [
```
Comment (upchuk) [inline src/Controller/ListPageRssController.php:216]: Let's key this array by the actual xml node element (pub date) in this case. And let's not call it `key` but `tag` no? Also, attributes needs to be an array. Matter of fact, why arent use using the `html_tag` render element here. That has all you need and you can just print it in your template without having to loop and check and stuff.

File: src/Controller/ListPageRssController.php [hunk 2]
```diff
@@ -164,12 +191,44 @@ public function build(NodeInterface $node): CacheableResponse {
...
+            'key' => 'pubDate',
+            'value' => $this->dateFormatter->format($creation_date, 'custom', DateTimeInterface::RFC822),
+            'attributes' => '',
+          ],
+        ],
+        '#entity' => $entity,
```
Comment (upchuk) [inline src/Controller/ListPageRssController.php:222]: Why is this here? It's not a variable of the theme hook

File: src/ListPageRssItemAlterEvent.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+declare(strict_types = 1);
+
+namespace Drupal\oe_list_pages;
+
+use Drupal\Core\Entity\EntityInterface;
+use Symfony\Component\EventDispatcher\Event;
+
+/**
+ * Event thrown in order to alter an RSS item build before its rendered.
```
Comment (upchuk) [inline src/ListPageRssItemAlterEvent.php:11]: it's rendered

File: src/ListPageRssItemAlterEvent.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+   * The render array for an RSS list item.
+   *
+   * @var array
+   */
+  protected $build;
+
+  /**
+   * The entity being rendered as an RSS item.
+   *
+   * @var \Drupal\Core\Entity\EntityInterface
```
Comment (upchuk) [inline src/ListPageRssItemAlterEvent.php:25]: ContentEntityInterface (everywhere)

**Review by upchuk — CHANGES_REQUESTED** [review]

File: src/Controller/ListPageRssController.php [hunk 5]
```diff
@@ -164,12 +192,57 @@ public function build(NodeInterface $node): CacheableResponse {
...
+      $entity = $item->getOriginalObject()->getEntity();
+      $cache_metadata->addCacheableDependency($entity);
+      $creation_date = $entity->get('changed')->value;
+      if ($entity->hasField('body')) {
+        $description = $entity->body->view([
+          'type' => 'text_default',
+          'label' => 'hidden',
+          'settings' => [],
+        ]);
+        $description = Html::escape($this->renderer->renderPlain($description[0]));
```
Comment (upchuk) [inline src/Controller/ListPageRssController.php:221]: Why you escaping here again? Twig already escapes no?

Seems we get the same result in Views so let's add a comment here as to why we are double escaping.

File: src/Controller/ListPageRssController.php [hunk 6]
```diff
@@ -164,12 +192,57 @@ public function build(NodeInterface $node): CacheableResponse {
...
+      }
+      $description = $description ?? '';
+      $result_item = [
+        '#theme' => 'oe_list_pages_rss_item',
+        '#title' => $entity->label(),
+        '#link' => $entity->toUrl('canonical', ['absolute' => TRUE]),
+        '#guid' => $entity->toUrl('canonical', ['absolute' => TRUE]),
+        '#description' => $description,
+        '#item_elements' => [
```
Comment (upchuk) [inline src/Controller/ListPageRssController.php:230]: You didn't key this array. So when overriding we cannot easily access it in the array in case we wanna change pubDate. See my other comment!

File: src/ListPageEvents.php [hunk 7]
```diff
@@ -24,10 +24,17 @@ final class ListPageEvents {
   const ALTER_BUNDLES = 'oe_list_pages.bundles_alter';
 
   /**
-   * Event fired when the RSS build before its rendered.
+   * Event fired after the RSS is built and before its rendered.
```
Comment (upchuk) [inline src/ListPageEvents.php:27]: it's

File: src/ListPageEvents.php [hunk 8]
```diff
@@ -24,10 +24,17 @@ final class ListPageEvents {
...
    *
    * @var string
    */
   const ALTER_RSS_BUILD = "oe_list_pages.rss_build_alter";
 
+  /**
+   * Event fired after the RSS item is built and before its rendered.
```
Comment (upchuk) [inline src/ListPageEvents.php:34]: it's

File: templates/oe-list-pages-rss-item.html.twig [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * @file
+ * Default theme implementation to display an item in a RSS feed.
+ *
+ * Available variables:
+ * - title: RSS item title.
+ * - link: RSS item link.
+ * - guid: RSS item global unique identifier.
+ * - description: RSS body text.
+ * - item_elements: RSS item elements to be rendered as XML (pubDate, creator,
```
Comment (upchuk) [inline templates/oe-list-pages-rss-item.html.twig:11]: isn't guid in a separate field? :P

File: tests/modules/oe_list_pages_event_subscriber_test/src/EventSubscriber/ListPagesTestSubscriber.php [hunk 10]
```diff
@@ -97,4 +99,22 @@ public function onRssBuildAlter(ListPageRssAlterEvent $event): void {
     $event->setBuild($build);
   }
 
+  /**
+   * Event handler for adding channel information.
```
Comment (upchuk) [inline tests/modules/oe_list_pages_event_subscriber_test/src/EventSubscriber/ListPagesTestSubscriber.php:103]: #doubt

File: tests/src/FunctionalJavascript/ListPageRssControllerTest.php [hunk 11]
```diff
@@ -90,10 +90,29 @@ public function testListPageRssPage() {
     $this->assertEquals('http://web:8080/build/core/themes/classy/logo.svg', $channel->filterXPath('//image/url')->text());
     $this->assertEquals('European Commission logo', $channel->filterXPath('//image/title')->text());
     $this->assertEquals('http://web:8080/build/', $channel->filterXPath('//image/link')->text());
-    // Assert modules subscribing to the ListPageRssBuildAlterEvent can
+    // Assert modules subscribing to the ListPageRssAlterEvent can
     // alter the build.
     $this->assertEquals('custom_value', $channel->filterXPath('//custom_tag')->text());
 
+    // Assert contents of items.
```
Comment (upchuk) [inline tests/src/FunctionalJavascript/ListPageRssControllerTest.php:97]: I think you forgot to assert the `creationDate` which you add in the subscriber

File: src/Controller/ListPageRssController.php [hunk 12]
```diff
@@ -164,12 +192,57 @@ public function build(NodeInterface $node): CacheableResponse {
...
+   * @param \Drupal\Core\Cache\CacheableMetadata $cache_metadata
+   *   The current cache metadata.
    *
    * @return array
    *   Array of items to be rendered.
    */
-  protected function getItemList(NodeInterface $node): array {
-    return [];
+  protected function getItemList(NodeInterface $node, CacheableMetadata $cache_metadata): array {
+    $configuration = ListPageConfiguration::fromEntity($node);
```
Comment (upchuk) [inline src/Controller/ListPageRssController.php:202]: We are missing the sorting: `Always sort by updated date, DESC. The RSS needs to show latest items.`

At the moment, we are sorting based on the bundle third party setting (see `executeList()`). So we need to make a ListPageConfiguration option where we can pass a different sort and use that if it's there instead of the content type.

**Review by upchuk — CHANGES_REQUESTED** [review]

File: src/Controller/ListPageRssController.php [hunk 13]
```diff
@@ -164,12 +192,57 @@ public function build(NodeInterface $node): CacheableResponse {
...
+    $execution_result = $this->listExecutionManager->executeList($configuration);
+    $query = $execution_result->getQuery();
+    $results = $execution_result->getResults();
+    $cache_metadata->addCacheableDependency($query);
+    $cache_metadata->addCacheableDependency($results);
```
[the remark this answers is not in the document]
Comment (imanoleguskiza) [reply src/Controller/ListPageRssController.php:207]: Thanks for the tip

File: src/Controller/ListPageRssController.php [hunk 14]
```diff
@@ -164,12 +192,68 @@ public function build(NodeInterface $node): CacheableResponse {
...
+      $cache_metadata->addCacheableDependency($entity);
+      $description = '';
+      // Add a default description if the body is available.
+      if ($entity->hasField('body')) {
+        $description = $entity->get('body')->view([
+          'type' => 'text_default',
+          'label' => 'hidden',
+          'settings' => [],
+        ]);
+        // We need to escape the results of renderPlain in order to maintain
```
Comment (upchuk) [inline src/Controller/ListPageRssController.php:226]: Is that what we are doing? p tags are not lost in twig. Twig autoesapes but leaves basic markup..

File: src/ListPageEvents.php (hunk 7 above)
Comment (upchuk) [inline src/ListPageEvents.php:27]: it's

File: tests/src/FunctionalJavascript/ListPageRssControllerTest.php [hunk 15]
```diff
@@ -48,39 +49,56 @@ public function testListPageRssPage() {
...
-    // Index the nodes.
     $index->indexItems();
 
     $node = $this->drupalGetNodeByTitle('List page test');
     $this->drupalLogout();
 
-    // Assert contents of channel elements.
+    // Assert the default sorting order of the list is
+    // by creation date, descending.
+    $this->drupalGet(Url::fromRoute('entity.node.canonical', ['node' => $node->id()]));
```
Comment (upchuk) [inline tests/src/FunctionalJavascript/ListPageRssControllerTest.php:86]: Not sure you need all this. You can simply just deduce the order below when you do things like `$first_item = $items->eq(0);`

File: src/Controller/ListPageRssController.php (hunk 12 above)
Comment (upchuk) [reply src/Controller/ListPageRssController.php:202]: Good job. The test also confirms it.

File: tests/src/FunctionalJavascript/ListPageRssControllerTest.php (hunk 15 above)
Comment (imanoleguskiza) [reply tests/src/FunctionalJavascript/ListPageRssControllerTest.php:86]: I'm first checking that the normal list is ordered by creation date and then through the other tests we assert that the rss list is ordered by the changed.

File: src/Controller/ListPageRssController.php [hunk 16]
```diff
@@ -164,12 +191,68 @@ public function build(NodeInterface $node): CacheableResponse {
...
-  protected function getItemList(NodeInterface $node): array {
-    return [];
+  protected function getItemList(NodeInterface $node, CacheableMetadata $cache_metadata): array {
+    $configuration = ListPageConfiguration::fromEntity($node);
+    // Always sort by the updated date descending.
+    $configuration->setSort([
+      'name' => 'changed',
+      'direction' => 'DESC',
+    ]);
+    $execution_result = $this->listExecutionManager->executeList($configuration);
```
Comment (hernani) [inline src/Controller/ListPageRssController.php:207]: There is requirement stating that limit for rss pages should be 25 items, we need to set it here and then test it. Currently default limit 10 is used.

File: tests/src/FunctionalJavascript/ListPageRssControllerTest.php (hunk 15 above)
Comment (upchuk) [reply tests/src/FunctionalJavascript/ListPageRssControllerTest.php:86]: @imanoleguskiza I know what you are doing. But below you are doing `$first_item = $items->eq(0);` which means you expect the first one (0) to be something. So with that you also assert the order cause you only have 2 items. You dont need a separate thing.

File: src/Controller/ListPageRssController.php (hunk 16 above)
Comment (upchuk) [reply src/Controller/ListPageRssController.php:207]: @imanoleguskiza Indeed, you can pass the limit to the configuration object.

**Review by upchuk — APPROVED** [review]

**Review by hernani — APPROVED** [review]
